external/vcm/vcm/limit.py

```diff
@@ -0,0 +1,90 @@
+import xarray as xr
+from typing import Sequence, Mapping, Union, Optional
+
+
+def _limit_extremes(data: xr.DataArray, limits: xr.DataArray) -> xr.DataArray:
+    """Limit values beyond prescribed extremes in a data array
+    
+    Args:
+        data: Data to be limited
+        limits: variable containing limits of the data, which must include a 'bounds'
+            dimension containing 'upper' and 'lower' coordinates. Other dimensions
+            should be a subset of the data dimensions.
+            
+    Returns: a limited data array
+    
+    """
+    upper = limits.sel(bounds="upper")
+    lower = limits.sel(bounds="lower")
+    return data.where(data < upper, upper).where(data > lower, lower)
+
+
+class LimitedDataset(xr.Dataset):
```

Comment:
Suggest turning this into a function returning an xarray object. the inheritance allows you to lazily apply the limiting, but we can do that with dask.

In general it’s not easy to subclass data structure like an xarray dataset. This goes back to the datastructure vs object stuff in the clean code book. Testing that this class is a valid dataset is nigh impossible…note how xarray itself doesn’t even have subclasses of dataset. There are simply too many methods/attributes to keep track of.

I also suggest  isolating this logic to the prognostic run? Or are you considering using it in fv3fit too for training?